[Tey Cheng Jun] iP

CJun1039 opened the pull request:
### Mr. Duke

> The next generation of AI companions. - Bob ([source](https://www.nus.edu.sg/))

- Strong
- Reliable
- Handsome

Get Mr. Duke **NOW**!

Instructions:

1. *Subscribe* to our mailing list
2. ~~Complete~~ CS2103T
3. [Try it for free!](https://www.nus.edu.sg/)

Taking care of Mr. Duke 💅

- [x] Talk to him daily
- [ ] Feed him food
- [ ] Take him out for walks

Check out this cool `HelloWorld` method

```Java
String HelloWorld() {
    System.out.println("Hello World!");
    return "Hello World";
}
String Hello = HelloWorld();
```

Commit:
Add Gradle support

CJun1039 committed:
add Greet, Echo, Exit to Duke.

CJun1039 committed:
add "add", list to Duke

CJun1039 committed:
add task and changed duke's list command to include the task. also allowed duke to mark and unmark tasks

CJun1039 committed:
Added ToDos, Events and Deadlines.

CJun1039 committed:
add Automated Text UI Testing

CJun1039 committed:
Add error handling to Duke and add -w option to FC in runtest.bat

CJun1039 committed:
add delete to duke

CJun1039 committed:
Add save to Duke

CJun1039 committed:
Merge branch 'branch-Level-7'

CJun1039 committed:
Revert "Add save to Duke"

This reverts commit 13b3d89973e762601aa8c021206ea5a83a4fd1fa.

CJun1039 committed:
Revert "Revert "Add save to Duke""

This reverts commit e5536c92c90a0e010bc80e304e2feea003ecd096.

CJun1039 committed:
Merge branch 'branch-Level-7'

CJun1039 committed:
Add date time formatting to Duke

CJun1039 committed:
Add more date time formatting to Duke

CJun1039 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Task.java

CJun1039 committed:
Add OOP to the implementation

CJun1039 committed:
Organize classes into package

CJun1039 committed:
Merge branch 'add-gradle-support' of https://github.com/CJun1039/ip into add-gradle-support

CJun1039 committed:
Add JUnit tests to duke

CJun1039 committed:
Improve exception handling and package duke into jar

CJun1039 committed:
add JavaDocs

CJun1039 committed:
Improve coding standard

CJun1039 committed:
Add find to duke

CJun1039 committed:
Merge branch 'branch-A-JavaDoc'

CJun1039 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java

CJun1039 committed:
Merge branch 'branch-Level-9'

chloeelim commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Constructs a Duke.
+     *
+     * @param filePath string representing the file path of a save file
```
I noticed this a few times throughout the code, perhaps there should be a punctuation after the `filePath` param description, and the first letter of the description should be capitalised? [(Reference: Punctuation behind each parameter description)](https://se-education.org/guides/conventions/java/index.html#comments)

```suggestion
     * @param filePath String representing the file path of a save file.
```

CJun1039 replied:
Thanks for helping me spot my mistakes!

chloeelim commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,36 @@
+    private final String message;
+
+    /**
+     * Constructs a DukeException
```
Should there be a punctuation after the Javadoc header description?

```suggestion
     * Constructs a DukeException.
 ```

chloeelim commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,36 @@
+ * DukeException represents the possible exceptions that might occur while using Duke.
+ */
+public class DukeException extends RuntimeException {
+    private final String message;
```
I think it would be better if constant names were in all uppercase? [(Reference: Constant names must be all uppercase using underscore to separate words)](https://se-education.org/guides/conventions/java/index.html#naming)

```suggestion
    private final String MESSAGE;
```

CJun1039 replied:
I agree with you!

chloeelim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+package duke;
+
+import duke.command.*;
```
Perhaps consider importing classes explicitly instead of using a wildcard import? [(Reference: Imported classes should always be listed explicitly)](https://se-education.org/guides/conventions/java/index.html#package-and-import-statements)

CJun1039 replied:
I used wildcard as I am using every class in that package. Perhaps I should create a new command parser within the command package.

chloeelim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+                    res = new DeleteCommand(Integer.parseInt(inputs[1]) - 1);
+                    break;
+                default:
+                    res = new AddCommand(inputs);
```
Should the case/ default clauses have the same indentation as the switch statement instead? [(Reference: The switch statement should have the following form: Note there is no indentation for case clauses)](https://se-education.org/guides/conventions/java/index.html#layout)

chloeelim commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+
+import duke.task.*;
+
+import java.io.*;
```
Perhaps you could consider importing the classes explicitly instead? [(Reference: Imported classes should always be listed explicitly)](https://se-education.org/guides/conventions/java/index.html#package-and-import-statements)

chloeelim commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+        } catch (IOException e) {
+            throw new DukeException(e.getMessage());
+        }
+        this.file.delete();
```
Maybe you shouldn't use the `this` keyword here since there are no method params shadowing the class variable `file`? [(Reference: Avoid unnecessary use of this with fields)](https://se-education.org/guides/conventions/java/index.html#variables)

```suggestion
       file.delete();
```

chloeelim commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+        }
+        this.file.delete();
+        temp.renameTo(new File("data/tasks.txt"));
+        this.file = temp;
```
Likewise for this line, maybe remove the `this` keyword?

chloeelim commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,88 @@
+        dukePrint("Something bad happened! :O", "Error message: " + msg);
+    }
+
+<<<<<<< HEAD
```
I think you forgot to delete this line when resolving your merge conflict?

chloeelim commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @return a String containing the user input
+     */
+=======
+>>>>>>> branch-A-CodingStandard
```
These two lines as well, I think you forgot to delete them when you did were resolving your merge conflict?

chloeelim commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Represents the command for the addition of new tasks to Duke's TaskList.
+ */
+public class AddCommand implements Command{
```
I think there's a missing white space before the open curly brace here?

```suggestion
public class AddCommand implements Command {
```

chloeelim commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public class DeleteCommand implements Command{
+
+    private final int toDelete;
```
Perhaps `toDelete` should be capitalised since it is a constant? [(Reference: Constant names must be all uppercase using underscore to separate words)](https://se-education.org/guides/conventions/java/index.html#naming)

```suggestion
    private final int TO_DELETE;
```

chloeelim commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Represents the command for asking Duke to list its tasks in the TaskList.
+ */
+public class ListCommand implements Command{
```
Should there be a whitespace before the open curly braces?

```suggestion
public class ListCommand implements Command {
```

chloeelim commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,45 @@
+ * Represents the command for marking a task in Duke's TaskList.
+ */
+public class MarkCommand implements Command{
+    private final int toMark;
```
Similarly, I think `toMark` should be capitalised since it is a constant? (There might be a few other instances of this styling error in other parts of the code)

chloeelim commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,91 @@
+package duke.task;
+
+<<<<<<< HEAD
+import java.time.LocalDate;
+
+/**
+ * An abstract class representing the Tasks included in Duke.
+ */
+=======
+>>>>>>> branch-A-CodingStandard
```
I think you forgot to delete these lines again when resolving your merge conflict?

```suggestion
import java.time.LocalDate;

/**
 * An abstract class representing the Tasks included in Duke.
 */
```

chloeelim commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    /**
+     * Format for printing messages.
```
Should this be pluralized (`formats` instead of `format`)?

chloeelim reviewed (commented):
Overall your code looks quite clean and readable, but there are some styling errors I noticed. Your Javadoc is quite extensive and descriptive, so good job on that! 👍 

Remember you don't accidentally commit the `<<<<<<< HEAD`, `=======` and `>>>>>>> branch-<branch merging>` lines when you resolve merge conflicts though, your code might have some trouble running if I'm not mistaken? 😅

CJun1039 committed:
Fix javadocs styling error

CJun1039 committed:
Improved java coding style.

AshiqurRah commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * Adds a new task to the save data.
```
I like that there are javadocs for every methods and constructors in all classes. Really commendable

AshiqurRah commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+     */
+    public static Command parse(String input) throws DukeException {
+        String[] inputs = input.split(" ", 2);
+        String head = inputs[0].toUpperCase();
```
head seems to be ambiguous name. One suggestion would be to change it to commandType

AshiqurRah reviewed (commented):
Looks good. Code is really easy to read

Yingming23 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+     * @return ArrayList of tasks retrieved from the save data.
+     */
+    public ArrayList<Task> load() {
+        ArrayList<Task> res = new ArrayList<>();
```
Perhaps use a different name such as list or taskList.

Yingming23 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Represents commands for duke to execute actions.
+ */
+public interface Command {
```
I like the idea to have Command be an interface.

Yingming23 reviewed (commented):
Overall, I like that your code is clean and OOP is implemented well. Well done.

CJun1039 committed:
improve coding standards

CJun1039 committed:
add more oop to duke

CJun1039 committed:
add gui to duke

CJun1039 committed:
Merge branch 'Level-10'

CJun1039 committed:
Update duke exception handling

DukeException:

Echoes java exception messages, unhelpful to the users.
One DukeException for all exceptions.

Add new DukeExceptions with messages helpful for the users.

CJun1039 committed:
Add assertions to duke

Assertions are suitable for verifying assumptions about Internal
Invariants, Control-Flow Invariants, Preconditions, Postconditions,
and Class Invariants.

Let's insert assertion statements liberally into the code to check the
assumptions on the program state during runtime.

CJun1039 committed:
Improve code quality of the source code

Some methods are long and difficult to read.

Split the methods into smaller methods and apply SLAP.

Let's split fileLineToTask() into a new class StorageReader and add
methods to read and parse the file lines in Storage.

Allows application of SLAP to increase code readability.

CJun1039 committed:
Merge pull request #1 from CJun1039/branch-A-Assertions

Add assertions to duke

CJun1039 committed:
Merge branch 'master' of https://github.com/CJun1039/ip

* 'master' of https://github.com/CJun1039/ip:
  Add assertions to duke

CJun1039 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java

CJun1039 committed:
Merge pull request #2 from CJun1039/branch-A-CodeQuality

Improve code quality of the source code

CJun1039 committed:
Merge branch 'master' of https://github.com/CJun1039/ip

CJun1039 committed:
Add tagging to duke

Users are categorizing tasks to events, todo, and deadline.

There can be more detail added to allow the users to further categorize
their tasks.

Let's add the ability to tag tasks to duke.

Each task can have a tag to it. During construction, a default empty
tag is assigned to the task object.

CJun1039 committed:
Merge branch 'branch-C-Tagging'

CJun1039 committed:
Add README.md to docs

CJun1039 committed:
Add and improve quality of JavaDocs

CJun1039 committed:
Improve duke GUI and add more personality to duke

CJun1039 committed:
Improve GUI and add personality to duke

CJun1039 committed:
Merge branch 'branch-A-BetterGui'

CJun1039 committed:
Add Ui.png to docs

CJun1039 committed:
Add more error handling to duke

CJun1039 committed:
Add checkstyle and fix codestyle issues

CJun1039 committed:
Change build.gradle configurations
